File: services/algolia_service.py
"""
Step 3 — Algolia Typo-Tolerant Inventory Search.

For every line flagged as requires_review, this service queries the Algolia
index using the official Python client (v4 async API).  Algolia's built-in
typo-tolerance handles the garbled OCR strings without any pre-processing.
"""

import asyncio
import logging
from typing import Any

from algoliasearch.search.client import SearchClient
from algoliasearch.search.models.search_params_object import SearchParamsObject

from config import settings
from models import AlgoliaMatch

logger = logging.getLogger(__name__)

# Module-level client — created once and reused across requests.
# The v4 async client is safe to share between coroutines.
_client: SearchClient | None = None


def _get_client() -> SearchClient:
    global _client
    if _client is None:
        _client = SearchClient(
            app_id=settings.ALGOLIA_APP_ID,
            api_key=settings.ALGOLIA_API_KEY,
        )
        logger.info("Algolia SearchClient initialised (app_id=%s).", settings.ALGOLIA_APP_ID)
    return _client


def _parse_hit(hit: Any) -> AlgoliaMatch:
    """
    Extract a clean AlgoliaMatch from a single Algolia Hit object.

    algoliasearch v4 uses Pydantic with extra="allow", so user-defined index
    fields (item_name, standard_quantity, …) are stored in hit.model_extra.
    Reserved fields (objectID → object_id) are first-class attributes.
    """
    # User-defined fields land in model_extra for Pydantic v2 extra="allow" models.
    attrs: dict[str, Any] = getattr(hit, "model_extra", {}) or {}

    object_id: str = getattr(hit, "object_id", "") or ""
    item_name: str = attrs.get("item_name", "")
    standard_quantity: str | None = attrs.get("standard_quantity")

    # Everything else passes through as metadata.
    metadata = {
        k: v for k, v in attrs.items()
        if k not in {"item_name", "standard_quantity"}
    }

    return AlgoliaMatch(
        object_id=object_id,
        matched_item_name=item_name,
        standard_quantity=standard_quantity,
        metadata=metadata,
    )


async def search_inventory(query_text: str) -> AlgoliaMatch | None:
    """
    Query the Algolia inventory index for *query_text*.

    Returns the top AlgoliaMatch, or None if there are no results or the
    request fails.  Algolia's typo-tolerance is enabled by default on all
    indices and requires no extra configuration here.
    """
    client = _get_client()

    params = SearchParamsObject(
        query=query_text,
        hits_per_page=1,        # we only need the best match
        attributes_to_retrieve=["item_name", "standard_quantity"],
        attributes_to_highlight=[],  # skip highlight markup — we don't need it
    )

    try:
        response = await client.search_single_index(
            index_name=settings.ALGOLIA_INDEX_NAME,
            search_params=params,
        )
    except Exception as exc:
        logger.error("Algolia search error for query '%s': %s", query_text, exc)
        return None

    hits = response.hits if response.hits else []
    if not hits:
        logger.warning("Algolia returned no hits for query: '%s'", query_text)
        return None

    match = _parse_hit(hits[0])
    logger.info(
        "Algolia matched '%s' → '%s' (objectID=%s)",
        query_text,
        match.matched_item_name,
        match.object_id,
    )
    return match


async def search_flagged_lines(flagged_texts: list[str]) -> list[AlgoliaMatch | None]:
    """
    Run search_inventory concurrently for every flagged line and return
    results in the same order as the input list.
    """
    tasks = [search_inventory(text) for text in flagged_texts]
    return list(await asyncio.gather(*tasks))

File: node-orchestrator/services/algoliaService.ts
import { algoliasearch } from "algoliasearch";
import crypto from "crypto";
import { AlgoliaMatch } from "../models.js";

let _client: ReturnType<typeof algoliasearch> | null = null;

function getClient() {
  if (!_client) {
    _client = algoliasearch(
      process.env.ALGOLIA_APP_ID!,
      process.env.ALGOLIA_API_KEY!
    );
    console.info(`Algolia client initialised (app_id=${process.env.ALGOLIA_APP_ID}).`);
  }
  return _client;
}

export async function searchInventory(
  queryText: string
): Promise<AlgoliaMatch | null> {
  try {
    const response = await getClient().searchSingleIndex({
      indexName: process.env.ALGOLIA_INDEX_NAME!,
      searchParams: {
        query: queryText,
        hitsPerPage: 1,
        attributesToRetrieve: ["item_name", "standard_quantity"],
        attributesToHighlight: [],
      },
    });

    if (!response.hits.length) {
      console.warn(`Algolia returned no hits for query: '${queryText}'`);
      return null;
    }

    const hit = response.hits[0] as Record<string, unknown> & {
      objectID: string;
    };
    const { objectID, item_name, standard_quantity, _highlightResult, ...rest } =
      hit;

    console.info(
      `Algolia matched '${queryText}' → '${item_name}' (objectID=${objectID})`
    );
    return {
      object_id: objectID,
      matched_item_name: (item_name as string) ?? "",
      standard_quantity: standard_quantity as string | undefined,
      metadata: rest as Record<string, unknown>,
    };
  } catch (err) {
    console.error(`Algolia search error for '${queryText}':`, err);
    return null;
  }
}

export async function searchFlaggedLines(
  flaggedTexts: string[]
): Promise<(AlgoliaMatch | null)[]> {
  return Promise.all(flaggedTexts.map(searchInventory));
}

export async function saveIngredients(
  ingredientNames: string[]
): Promise<number> {
  if (!ingredientNames.length) return 0;

  const seen = new Set<string>();
  const objects: object[] = [];

  for (const name of ingredientNames) {
    const normalised = name.toLowerCase().trim();
    if (!normalised || seen.has(normalised)) continue;
    seen.add(normalised);
    const objectID = crypto
      .createHash("md5")
      .update(normalised)
      .digest("hex");
    objects.push({ objectID, item_name: name.trim() });
  }

  if (!objects.length) return 0;

  await getClient().saveObjects({
    indexName: process.env.ALGOLIA_INDEX_NAME!,
    objects,
  });

  console.info(
    `Upserted ${objects.length} ingredient(s) to Algolia index '${process.env.ALGOLIA_INDEX_NAME}'.`
  );
  return objects.length;
}
